src/panels/calendar/ha-panel-calendar.ts

```diff
@@ -0,0 +1,281 @@
+            <ha-button-menu .icon=${"hass:dots-vertical"}>
+              <mwc-list-item @click=${this._handleRefresh}
+                >Refresh</mwc-list-item
+              >
+            </ha-button-menu>
+          </app-toolbar>
+        </app-header>
+        <div class="content">
+          <div class="calendar-list">
+            <div class="calendar-list-header">My Calendars</div>
+            ${this._calendars.map(
+              (selCal) =>
+                html`<div>
+                  <mwc-formfield .label=${selCal.calendar.name}>
+                    <mwc-checkbox
+                      style=${styleMap({
+                        "--mdc-theme-secondary":
+                          selCal.calendar.backgroundColor,
+                      })}
+                      .value=${selCal.calendar.entity_id}
+                      .checked=${selCal.selected}
+                      @change=${this._handleToggle}
+                    ></mwc-checkbox>
+                  </mwc-formfield>
+                </div>`
+            )}
+          </div>
+          <ha-full-calendar
+            .events=${this._events}
+            .narrow=${this.narrow}
+            @view-changed=${this._handleViewChanged}
+          ></ha-full-calendar>
+        </div>
+      </app-header-layout>
+    `;
+  }
+
+  private get _selectedCalendars(): Calendar[] {
+    return this._calendars
+      .filter((selCal) => selCal.selected)
+      .map((cal) => cal.calendar);
+  }
+
+  private async _fetchEvents(start: Date, end: Date, calendars: Calendar[]) {
+    if (!calendars.length) {
+      return;
+    }
+
+    const events = await fetchCalendarEvents(this.hass, start, end, calendars);
+
+    this._events = [...this._events, ...events];
+  }
+
+  private _handleToggle(ev) {
+    this._calendars = this._calendars.map((cal) => {
+      if (ev.target.value !== cal.calendar.entity_id) {
+        return cal;
+      }
+
+      const checked = ev.target.checked;
+
+      if (checked) {
+        this._fetchEvents(this._start!, this._end!, [cal.calendar]);
+      } else {
+        this._events = this._events.filter(
+          (event) => event.calendar !== cal.calendar.entity_id
+        );
+      }
+
+      cal.selected = ev.target.checked;
+      return cal;
+    });
+  }
+
+  private _handleViewChanged(ev: HASSDomEvent<CalendarViewChanged>) {
+    const viewStart = ev.detail.start;
+    const viewEnd = ev.detail.end;
+
+    if (
+      this._start &&
+      this._end &&
+      this._start <= viewStart &&
+      this._end >= viewEnd
+    ) {
+      return;
+    }
+
+    if (!this._start || !this._end) {
+      this._start = viewStart;
+      this._end = viewEnd;
+      this._fetchEvents(this._start, this._end, this._selectedCalendars);
+      return;
+    }
+
+    // If the date range moved to the left
+    if (viewStart < this._start && viewEnd >= this._start) {
+      this._fetchEvents(viewStart, this._start, this._selectedCalendars);
```

Comment:
What is the new end bigger than the old end? Like when I go from a day to a month view?